### PR TITLE
Replace User Information List usage with role assignments for SPO sites

### PR DESCRIPTION
**Pull request by seanstory**
## Closes https://github.com/elastic/connectors-python/issues/1651

- [x] uses role assignments for sites
- [x] uses consistent group-vs-user detection throughout
- [x] grabs both `grantedToV2` AND `grantedToIdentitiesV2`
- [x] fetches loginName where applicable
- [x] grabs groupIds instead of group emails
- [x] always stores `user_id` for ACL document params
  - https://github.com/elastic/connectors-python/issues/1678
- [x] refactors variable names for consistency
- [x] propagates ACLs to subsites

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

## Release Note

Fixes an issue where SPO DLS was based on User Information Lists, rather than Role Assignments

### CODE REVIEW

File: connectors/sources/sharepoint_online.py [hunk 1]
```diff
@@ -1352,43 +1354,12 @@ def _is_site_admin(user):
...
 
-        async for user_information in self.client.user_information_list(site["id"]):
-            user = user_information["fields"]
-
+        async for role_assignment in self.client.site_role_assignments(site['webUrl']):
             user_access_control = set()
+            user_access_control.update(await self._get_access_control_from_role_assignment(role_assignment))
+            member = role_assignment["Member"]
 
-            if is_domain_group(user):
```
Comment (seanstory) [inline connectors/sources/sharepoint_online.py:1360]: probably need to check to make sure this didn't leave unused methods that can be cleaned up

File: connectors/sources/sharepoint_online.py [hunk 2]
```diff
@@ -2077,22 +2024,36 @@ async def _get_access_control_from_role_assignment(self, role_assignment):
...
 
-        def _access_control_for_user(user_):
-            user_access_control = []
+        def _has_limited_access(role_assignment):
+            bindings = role_assignment.get("RoleDefinitionBindings", [])
 
-            user_principal_name = user_.get("UserPrincipalName")
-            login_name = _get_login_name(user_.get("LoginName"))
+            # If there is no permission information, default to restrict access
+            if not bindings:
```
Comment (timgrein) [inline connectors/sources/sharepoint_online.py:2031]:
```suggestion
            if not bindings:
              self._logger.debug("No RoleDefinitionBindings found for INSERT SOME ROLE ID HERE")
```

File: connectors/sources/sharepoint_online.py [hunk 3]
```diff
@@ -82,6 +82,15 @@
 DELTA_NEXT_LINK_KEY = "@odata.nextLink"
 DELTA_LINK_KEY = "@odata.deltaLink"
 
+# See https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core/Model/SharePoint/Core/Public/Enums/PermissionKind.cs
```
Comment (timgrein) [inline connectors/sources/sharepoint_online.py:85]:
```suggestion
# See https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core/Model/SharePoint/Core/Public/Enums/PermissionKind.cs
# View items in lists, documents in document libraries, and Web discussion comments.
```

File: connectors/sources/sharepoint_online.py [hunk 4]
```diff
@@ -82,6 +82,15 @@
...
+VIEW_ITEM_MASK = 0x1
+
+# See https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core/Model/SharePoint/Core/Public/Enums/PermissionKind.cs
```
Comment (timgrein) [inline connectors/sources/sharepoint_online.py:88]:
```suggestion
# See https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core/Model/SharePoint/Core/Public/Enums/PermissionKind.cs
# View pages in a Site.
```

File: connectors/sources/sharepoint_online.py [hunk 5]
```diff
@@ -82,6 +82,15 @@
...
+VIEW_PAGE_MASK = 0x20000
+
+# See https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core/Model/SharePoint/Core/Public/Enums/RoleType.cs
+VIEW_ROLE_TYPES = [2, 3, 4, 5, 6, 7, 0xFF]
```
Comment (timgrein) [inline connectors/sources/sharepoint_online.py:92]:
```suggestion
READER = 2
CONTRIBUTOR = 3
WEB_DESIGNER = 4
ADMINISTRATOR = 5
EDITOR = 6
REVIEWER = 7
SYSTEM = 0xFF
VIEW_ROLE_TYPES = [READER, CONTRIBUTOR, WEB_DESIGNER, ADMINISTRATOR, EDITOR, REVIEWER, SYSTEM]
```

Rationale: https://refactoring.guru/replace-magic-number-with-symbolic-constant
We could also add a comment, but nothing blocking ofc.

**Review by timgrein — APPROVED** [review]
LGTM, I've ran some tests locally and didn't find anything obvious wrong. I'll recheck that against the new test instance with more test data, too. Only suggesting two more log messages and some explanatory comments, which would helped (me at least) to understand some things a bit faster, feel free to add or skip them.

Nice work 👏

**Review by timgrein — DISMISSED** [review]

File: connectors/sources/sharepoint_online.py [hunk 6]
```diff
@@ -334,9 +360,11 @@ async def fetch(self, url):
         return await self._get_json(url)
 
     async def post(self, url, payload):
-        self._logger.debug(f"Post to url: {url}")
+        self._logger.debug(f"Post to url: '{url}' with body: {payload}")
         async with self._post(url, payload) as resp:
-            return await resp.json()
+            result = await resp.json()
+            # self._logger.debug(f"Response was: {result}")
```
Comment (vidok) [inline connectors/sources/sharepoint_online.py:366]: I guess this line is not needed.

File: connectors/sources/sharepoint_online.py [hunk 7]
```diff
@@ -1450,8 +1418,8 @@ async def _user_access_control_doc(self, user):
 
         expanded_member_groups = user.get("transitiveMemberOf", [])
         if (
-            len(expanded_member_groups) < 100
-        ):  # $expand param has a max of 100: see: https://learn.microsoft.com/en-us/graph/known-issues#query-parameters
+            len(expanded_member_groups) < 20
```
Comment (vidok) [inline connectors/sources/sharepoint_online.py:1421]: nit: it'd be nice to convert it to a constant.

**Review by vidok — APPROVED** [review]
Left some non-blocking comments. 


LGTM!

**Comment by seanstory** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|pre-8.10-stable|[](https://github.com/elastic/connectors-python/pull/1711)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by seanstory** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.10|[](https://github.com/elastic/connectors-python/pull/1713)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Pull request merged**